synapse/handlers/e2e_keys.py

```diff
@@ -976,15 +980,63 @@ def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None
 
         Raises:
             NotFoundError: if the key is not found
+            SynapseError: if `user_id` is invalid
         """
+        user = UserID.from_string(user_id)
         key = yield self.store.get_e2e_cross_signing_key(
-            user_id, key_type, from_user_id
+            user_id, desired_key_type, from_user_id
         )
+
+        # If we still can't find the key, and we're looking for keys of another user,
+        # then attempt to fetch the missing key from the remote user's server.
+        #
+        # We may run into this in possible edge cases where a user tries to
+        # cross-sign a remote user, but does not share any rooms with them yet.
+        # Thus, we would not have their key list yet. We fetch the key here and
+        # store it just in case.
+        if (
+            key is None
+            and not self.is_mine(user)
+            # We don't get "user_signing" keys from remote servers, so disallow that here
+            and desired_key_type != "user_signing"
```

Comment:
Might be better to list the key types that we do expect, rather than the one we don't, so that we drop `desired_key_type == "foo"` instead of uselessly trying to fetch it.